codegen/service/service_data.go

```diff
@@ -597,3 +743,242 @@ func requirements(m *design.MethodExpr) []*design.SecurityExpr {
 	}
 	return design.Root.API.Requirements
 }
+
+// isProjectable returns true if the given type is a result type with
+// more than one view and not a collection.
+func isProjectable(dt design.DataType) bool {
+	if design.IsArray(dt) {
+		return false
+	}
+	rt, ok := dt.(*design.ResultTypeExpr)
+	return ok && len(rt.Views) > 1
+}
+
+// buildViewedType builds the type data to generate views and the constructor
+// code to transform result type to its viewed counterpart (and vice versa).
+func buildViewedType(vAtt, at *design.AttributeExpr, scope *codegen.NameScope, viewsPkg string) *ViewedTypeData {
+	resultVar := scope.GoTypeName(at)
+	ut := vAtt.Type.(design.UserType)
+	dt := at.Type.(design.UserType)
+	if !isProjectable(dt) {
+		varName := scope.GoTypeName(vAtt)
+		return &ViewedTypeData{
+			UserTypeData: &UserTypeData{
+				Name:        ut.Name(),
+				Description: fmt.Sprintf("%s is the transformed type of %s type.", varName, resultVar),
+				VarName:     varName,
+				Def:         scope.GoTypeDef(ut.Attribute(), true),
+				Ref:         scope.GoTypeRef(vAtt),
+				Type:        ut,
+			},
+			Validate: codegen.RecursiveValidationCode(ut.Attribute(), true, false, false, "result"),
+		}
+	}
+	vt := design.Dup(ut).(design.UserType)
+	vt.Rename(ut.Name() + "View")
+	vAtt.Type = vt
+	ref := scope.GoTypeRef(vAtt)
+	varName := codegen.Goify(vt.Name(), true)
+	return &ViewedTypeData{
+		ViewType: &UserTypeData{
+			Name:        vt.Name(),
+			Description: fmt.Sprintf("%s is the transformed type of %s type.", varName, resultVar),
```

Comment:
Not sure "transformed" is going to make sense to users. Probably need more explanation here.